WIP: Add a Patchwork interface to PaStA

metp opened the pull request:
WIP: Add setup instructions

closes #9

metp committed:
Move remove_identical_eq_classes to Clustering

Signed-off-by: Mete Polat <metepolat2000@gmail.com>

metp committed:
Automate pasta/patchwork workflow

Signed-off-by: Mete Polat <metepolat2000@gmail.com>

bulwahn reviewed (approved):
Looks good. We might actually consider to have this Dockerfile located and referenced somewhere more general than docker/patchwork/, as this is not only useful for the patchwork setup but for other setups as well.

bulwahn reviewed (commented):
Commit "Add function to retrieve latest message date" looks sound from the technical point, but the commit message should explain why you actually need this function for the patchwork integration.

bulwahn reviewed (commented):
Commit Add integration for pulling patches from patchwork:
using "pw" looks like a bad choice for an abbreviation of the argument option (pw reminds too much of password, not patchwork), either go with the full string "patchwork" or simply use "-p" (if not already used).

bulwahn reviewed (commented):
Commit "Skip download of already available mbox" looks good.

bulwahn reviewed (commented):
Commit "Move remove_identical_eq_classes to Clustering" looks like a reasonable refactoring change. This could be already merged independently before.

bulwahn reviewed (commented):
The great magic happening in "Add push cmd for pushing relations to patchwork" looks good to me. I would love to see that in action in a small productive patchwork instance :)

bulwahn commented:
@rralf We are waiting for your review. I hope we can get this merged in the pasta master in the four weeks and then reach out on the patchwork mailing list to have some alpha users trying to use this feature.
@Honeybyte Do you think we are ready in four weeks for the first alpha users?

bulwahn commented:
I am wondering if @Honeybyte should provide an email address under which he is reachable more long-term than the current one provided. I think I would prefer that. Copyright is already clarified by the header statement.

metp commented:
> Looks good. We might actually consider to have this Dockerfile located and referenced somewhere more general than docker/patchwork/, as this is not only useful for the patchwork setup but for other setups as well.

You are right. The main difference between existing Dockerfiles and the one in the patchwork subdirectory is that the latter one is more suitable for a development environment as the source code lives on the host machine rather than in the container itself. I propose naming the directory 'development' instead.

metp commented:
> I am wondering if @Honeybyte should provide an email address under which he is reachable more long-term than the current one provided. I think I would prefer that. Copyright is already clarified by the header statement.

Yes sure, sounds reasonable. I will change the signed-off-by address.

metp commented:
> @Honeybyte Do you think we are ready in four weeks for the first alpha users?

We are ready now. The code on the Patchwork side is also finished (just some non source code things like edit docs, etc. is missing). Same on pasta side. Only setup instructions are missing.

metp commented:
> The great magic happening in "Add push cmd for pushing relations to patchwork" looks good to me. I would love to see that in action in a small productive patchwork instance :)

https://drive.google.com/file/d/1nkto3kIk1xK65MCgmt_rcMcYc3_Mp0Gi/view?usp=sharing

bulwahn commented:
Nice video...
@rralf is going to review this pull request next week.
Once the patch set is provided on the mailing list and installation instructions are written, could we set up an own "shadow" lkml patchwork instance with the pasta integration running?
E.g., obtain a "lkml mbox dump" from lore.kernel.org and set up patchwork with that data, then subscribe to lkml and let the patchwork instance follow the mailing list, have that patchwork instance connected to pasta and show the continuous patchwork-pasta integration running. That would make a great demo for ELCE...

rralf commented:
Sorry for the delay... I'm about to review the patches -- is this the latest version? Thanks.

rralf commented on docker/development/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+# Timezone
+ARG TZ="Europe/Berlin"
+# Configure tzdata
+RUN rm /etc/localtime; ln -s /usr/share/zoneinfo/$TZ /etc/localtime
```
why don't you use timedatectl/localectl?

metp replied:
They are part of systemd which isn't included in the ubuntu docker image as there is no need for an init system within a container.

rralf replied:
Hm. I see. Instead of using Europe/Berlin, I'd rather prefer to use the host's tz. Can we simply copy it? Could be patched in pasta-base.dockerfile.

rralf replied:
I investigated how others solve this issue. Conclusion is that timezone information is something that shouldn't be set during container build-time, it should be set during runtime of the container. One way to do that is, e.g., docker run -v /etc/localtime:/etc/localtime:ro . You could also define your own local dockerfile overlay that does timezone stuff at the end.

rralf commented on docker/development/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+RUN pip3 install \
+    anytree \
+    flask-bootstrap \
+    flask-nav
```
You are duplicating things. Please inherit from pasta-base.dockerfile or pastalinux.dockerfile.

rralf replied:
Would that help? https://github.com/lfd/PaStA/commit/967a09ade46c0cc883143cc109f0fee9ab0a8097

metp replied:
Yes that works, thank you 👍

metp commented:
> Sorry for the delay... I'm about to review the patches -- is this the latest version? Thanks.

Yes

rralf reviewed (changes requested):
Please use pasta-{base,linux} if possible and don't introduce dockerfiles from scratch.

rralf reviewed (commented):
Is already forwarded on master, please rebase to the latest next branch.

rralf reviewed (commented):
No need for implementing a non-interactive mode as e already have it: Simply set the auto-accept threshold lower than the interactive threshold.

rralf commented on pypasta/Config.py:
```diff
@@ -195,6 +196,10 @@ def path(name):
 
+            mbox_pw = 'patchwork'
+            if mbox_pw in mbox:
+                self.pw = Patchwork(mbox[mbox_pw]['url'])
```
These three lines will be completely modified two patches above, can we somehow use the eventual version of this change in this patch?

rralf replied:
Why does the instance need to be a member of Config? Wouldn't the mbox subsystem (e.g., repo.mbox.patchwork) be a better place?

rralf commented on pypasta/Config.py:
```diff
@@ -195,6 +196,10 @@ def path(name):
             mbox_raw = mbox['raw']
             mbox_pub_in = mbox['pubin']
 
+            mbox_pw = 'patchwork'
```
What does mbox_pw mean?

rralf commented on pypasta/Config.py:
```diff
@@ -299,6 +304,28 @@ def load_upstream_hashes(self, force_reload=False):
+                pulled += 1
+        finally:
+            for f_mbox in [file for file in lists.values()
+                           if not isinstance(file, str)]:
```
Can we somehow avoid the isinstance dance? isinstance always makes the assumption that the datatype might change. This is error prone.

metp replied:
The datatype is changing here. What is happening here is that we append a patch to its corresponding list after downloading it. The path to the list file is replaced with an actual file object then.
We can still avoid using isinstance of course.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -454,6 +457,19 @@ def update(self, nofetch):
+    def latest_message_date(self):
+        latest = datetime.datetime.min
+        for mbox in itertools.chain(self.pub_in, self.mbox_raw.raw_mboxes):
+            f_index = mbox.f_index if isinstance(mbox, PubInbox) \
```
Please no isinstance(). Just implement helper functions in both classes that do the job

metp replied:
Unfortunately helper functions wouldn't really "help" here. We want the index for a specific mbox. While we can use f_index directly on a pub_in mbox we cannot use something similar on a RawMbox as there is just one RawMbox for multiple lists.

rralf reviewed (commented):
Can we really rely on the latest message date? What if patchwork receives a deferred message from yesterday for some reason? What if a Mail has broken Date headers? Do we loose mails in this case?

rralf commented on pypasta/Config.py:
```diff
@@ -196,9 +196,9 @@ def path(name):
-                self.pw = Patchwork(mbox[mbox_pw]['url'])
+            mbox_pw = mbox.get('patchwork')
+            if mbox_pw:
+                self.pw = Patchwork(**mbox_pw)
```
This patch should be squashed with the patch that introduces the Patchwork class.

rralf commented on pypasta/Repository/Mbox.py:
```diff
@@ -316,8 +316,10 @@ def update(self):
 
-    def get_patchwork_ids(self, message_id):
-        return [cached[3] for cached in self.index[message_id]]
+    def get_patchwork_id(self, message_id):
```
Please don't introduce code and entirely remove it in the next patch. It's hard for me to review what your code does, when the semantics of the routines changes every other patch.

rralf reviewed (commented):
Looks good, but I need have to test this.

rralf commented on tools/pasta_pw:
```diff
@@ -0,0 +1,14 @@
+./pasta rate -ni
+
+# Upload patch relations and commit references on Patchowrk
+./pasta push
```
nit: missing newline at the end + license header

rralf replied:
Ah okay, now I understand what pasta push is doing...

rralf commented on tools/pasta_pw:
```diff
@@ -0,0 +1,14 @@
+set -e
+
+# Download new patches from Patchwork and add them to the right mboxes
+./pasta sync -mbox -p -nofetch
```
PASTA=./pasta

$pasta sync ...

rralf commented on tools/pasta_pw:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+# Stop script when one of the commands fails
+set -e
```
good idea ;-)

rralf reviewed (commented):
Can be squashed to another patch. Contains only changes to an other patch of this series.

rralf reviewed (commented):
Fixup of another patch

rralf reviewed (commented):
Fixup

rralf commented:
Hi Mete,

thanks for the patches!

I will need to test this series. You already created a dockerfile, though I don't understand what it is currently used for.

Can you extend this dockerfile to install pasta + a working patchwork instance? Without having a working patchwork / pasta instance, it's hard for me to test the changes.

Thanks
  Ralf

rralf commented:
> Commit "Move remove_identical_eq_classes to Clustering" looks like a reasonable refactoring change. This could be already merged independently before.

Ack, I will test that patch and consider that.

metp commented:
> Please use pasta-{base,linux} if possible and don't introduce dockerfiles from scratch.

I am not entirely happy with using pasta-base. As mentioned above, in comparison to pasta-base, the newly introduced Dockerfile does not copy anything into the container. Instead the source code can be mounted into a container allowing to skip rebuilding the image whenever the code changes.
What do think about:

1. Move all Python dependencies into a requirements.txt
2. Use the new Dockerfile as pasta-base and install all python dependencies from requirements.txt
3. Use the current pasta-base as a production Dockerfile (and the new Dockerfile as its base)
4. Base pasta-linux on production

rralf commented:
Fair enough. A three-stage setup would be fine for me. Something like base/pasta-base/pasta-linux.

Hmm. I never thought of mounting PaStA into the countainer. I'm curious: what does your workflow look like? (Nevertheless, for integration testing, we would still need to clone pasta in the container)

metp commented:
> Can we really rely on the latest message date? What if patchwork receives a deferred message from yesterday for some reason? What if a Mail has broken Date headers? Do we loose mails in this case?

I haven't thought about that, we could loose some mails. Fortunately Patchwork also exposes an events API which we can use instead.

metp committed:
Add python-requests as dependency

Used for HTTP requests to Patchwork.

Signed-off-by: Mete Polat <metepolat2000@gmail.com>

metp commented:
> Hmm. I never thought of mounting PaStA into the countainer. I'm curious: what does your workflow look like?

start patchwork (and create patchwork_default network): 
`$ docker-compose up`

build container without sending any context:
`$ docker build  -t pasta:latest - < docker/pasta-skeleton.dockerfile`

start pasta in the same network and connect to the container:
`$ docker run -it --rm --network patchwork_default --name pasta -v <host/path/to/PaStA>:/home/pasta pasta:latest`

Mounting the source code (with `-v`) into the container allows using the host source code in the container without needing to rebuild when changing code.

metp committed:
Add patchwork interface

Add the connection between PaStA and Patchwork. Patches can be
downloaded from Patchwork and be analysed in PaStA. The results can be
uploaded to Patchwork again.

New option in pasta_sync: -p
Download all or the most recent patches from Patchwork and include them
in PaStA.

New command: pasta_push
Upload related patches metadata and upstream commit references to a
Patchwork instance.

Configuration parameters for Patchwork:
- Url to a Patchwork instance
- Authentication credentials

Signed-off-by: Mete Polat <metepolat2000@gmail.com>

rsarky commented:
Hey!
What are the pending work items here?
I am considering picking Patchwork integration for PaStA as my GSoC project

rralf commented:
Just to name some examples:
  - Automatic updates of resources (e.g., tools/update.sh)
  - Regular, automated analyses of incoming patches/commits (docker, cron-jobs, ...).
  - Automated commits to the resources-repo, ...
  - Mapping of message-ids to patchwork-ids (Afair, Mete already did some work)
  - Submission of results to patchwork (can be installed via docker)
  - Traceability of analyses: which comparisons happened when, why did pasta take decisions, ...

rsarky commented:
Thanks @rralf . This helps. although the final goal: integrate PaStA with an instance of Patchwork is well defined, this is slightly tricky as this is already WIP. I wouldn't want to duplicate any work or go over things that have already been looked at. @Honeybyte Can you help figuring out some independent work items? 

For eg.
Submission of results to patchwork sounds like a good work item although as far as I can see Mete has already looked into this.

metp commented:
Hi @rsarky, happy to hear you want to contribute! I am currently not working anymore on combining pasta with patchwork so this task can definitely picked up by someone else. Patchwork recently integrated the required infrastructure for patch relations. The [REST relations API](https://github.com/getpatchwork/patchwork/commit/83f364aad66c35e31df8d0871ec2b62016eba337) in patchwork changed a lot so parts of this PR (including the submission of results routine) have to be updated.

rsarky commented:
Thanks for getting back Mete! :smile: 
I will go over this PR once to gauge what needs to be picked up and maybe make a rough list of items.
Will look into this first thing tomorrow morning.

rsarky commented:
Going over the PR and the list mentioned by @rralf, I can identify the following broad action points:
- Automate fetching patches from patchwork. @Honeybyte has already written a script [here](https://github.com/lfd/PaStA/pull/31/files#diff-fd0b2645929d8c7e4b325b473b8f6458) and as @rralf mentioned tools/update.sh can also be referred to. Will need to setup a cron job that automatically runs this script and calls PaStA APIs. Also automating commits to the resources repo can be made a part of this cron job too.
- Submitting relations to Patchwork. As @Honeybyte mentioned, the API has changes since his work on this PR and that will need to change. Also now that we have [support](https://github.com/lfd/PaStA/commit/2f19b3e483725804fcc11a67c1d082d7a01a71b0) to parse "Link:" tags referrring to (lore|lkml).kernel.org we should add those to the relations as well.
- Traceability of analyses. At this point I am not really sure how to go about this, but we basically need to figure out how to audit the decisions that PaStA has taken.

@Honeybyte , @bulwahn , @rralf : Can you think of anything that I have missed or gotten wrong?

metp commented:
Hi @rsarky, also have a look into https://github.com/lfd/PaStA-resources/pull/1. Beside that I cannot think of anything else. @rralf Can @rsarky push to the PRs when you assign him for them? Otherwise it's probably time now to close both PRs so that you can open your own ones (perhaps already a draft PR)

rralf commented:
I doubt that someone can "push" to a PR. The basis of this PR is your private fork, and I think that it won't change anything if I assign this issue to Rohit, but let me try. With those fancy web interfaces you never know ;-)

rsarky commented:
Hmm, not sure if I can check this PR out locally from here. I will probable need to set @Honeybyte 's fork as one of my upstreams in local and pull the branch from there.

rralf commented:
Yes, that's definitely the case.

rsarky commented:
Hey! Facing some issues building the pasta dockerfile.
Specifically, when running `docker build -t pasta:latest - < docker/pasta-skeleton.dockerfile` .
Here are the logs:
```
Sending build context to Docker daemon  3.072kB
Step 1/11 : FROM ubuntu:19.04
19.04: Pulling from library/ubuntu
4dc9c2fff018: Pull complete
0a4ccbb24215: Pull complete
c0f243bc6706: Pull complete
5ff1eaecba77: Pull complete
Digest: sha256:2adeae829bf27a3399a0e7db8ae38d5adb89bcaf1bbef378240bc0e6724e8344
Status: Downloaded newer image for ubuntu:19.04
 ---> c88ac1f841b7
Step 2/11 : MAINTAINER Ralf Ramsauer "ralf.ramsauer@oth-regensburg.de"
 ---> Running in 7c14ced68c17
Removing intermediate container 7c14ced68c17
 ---> 17892686dbb1
Step 3/11 : ENV DEBIAN_FRONTEND noninteractive
 ---> Running in 7666e0747c76
Removing intermediate container 7666e0747c76
 ---> bcfeed8568c5
Step 4/11 : ENV LANG="C.UTF-8"
 ---> Running in 412f9aeb523d
Removing intermediate container 412f9aeb523d
 ---> 3a881e8078ac
Step 5/11 : ENV LC_ALL="C.UTF-8"
 ---> Running in 1b1f3ea8e653
Removing intermediate container 1b1f3ea8e653
 ---> cde20e8830a3
Step 6/11 : RUN apt update && apt -y dist-upgrade
 ---> Running in 7e8c07d51383

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Ign:1 http://archive.ubuntu.com/ubuntu disco InRelease
Ign:2 http://security.ubuntu.com/ubuntu disco-security InRelease
Ign:3 http://archive.ubuntu.com/ubuntu disco-updates InRelease
Ign:4 http://archive.ubuntu.com/ubuntu disco-backports InRelease
Err:5 http://security.ubuntu.com/ubuntu disco-security Release
  404  Not Found [IP: 91.189.91.39 80]
Err:6 http://archive.ubuntu.com/ubuntu disco Release
  404  Not Found [IP: 91.189.88.142 80]
Err:7 http://archive.ubuntu.com/ubuntu disco-updates Release
  404  Not Found [IP: 91.189.88.142 80]
Err:8 http://archive.ubuntu.com/ubuntu disco-backports Release
  404  Not Found [IP: 91.189.88.142 80]
Reading package lists...
E: The repository 'http://security.ubuntu.com/ubuntu disco-security Release' does not have a Release file.
E: The repository 'http://archive.ubuntu.com/ubuntu disco Release' does not have a Release file.
E: The repository 'http://archive.ubuntu.com/ubuntu disco-updates Release' does not have a Release file.
E: The repository 'http://archive.ubuntu.com/ubuntu disco-backports Release' does not have a Release file.
The command '/bin/sh -c apt update && apt -y dist-upgrade' returned a non-zero code: 100
```

Wonder if you have seen this error before?

UPDATE: Fixed! see below

rsarky commented:
UPDATE: Changing ubuntu verison from `19.04` to `18.04` solved the issue!

rsarky commented:
> > Hmm. I never thought of mounting PaStA into the countainer. I'm curious: what does your workflow look like?
> 
> start patchwork (and create patchwork_default network):
> `$ docker-compose up`
> 
> build container without sending any context:
> `$ docker build -t pasta:latest - < docker/pasta-skeleton.dockerfile`
> 
> start pasta in the same network and connect to the container:
> `$ docker run -it --rm --network patchwork_default --name pasta -v <host/path/to/PaStA>:/home/pasta pasta:latest`
> 
> Mounting the source code (with `-v`) into the container allows using the host source code in the container without needing to rebuild when changing code.

Hey @Honeybyte ,
Any steps that you follow after this for testing?
I am getting some weird issue where PaStA cannot parse the `default.cfg`

rralf commented:
> UPDATE: Changing ubuntu verison from `19.04` to `18.04` solved the issue!

... You had to downgrade your local ubuntu version to get it working?

rralf commented:
>  I am getting some weird issue where PaStA cannot parse the `default.cfg`

I guess you will have to rebase this series to master/next. E.g., Rohit, you just changed the format of mbox-result -> patch-groups. This might explain the break...

rsarky commented:
> ... You had to downgrade your local ubuntu version to get it working?

No no, just the one in the `pasta-skeleton` dockerfile

rsarky commented:
> > I am getting some weird issue where PaStA cannot parse the `default.cfg`
> 
> I guess you will have to rebase this series to master/next. E.g., Rohit, you just changed the format of mbox-result -> patch-groups. This might explain the break...

Ah, that's possible, will try that

rralf commented:
> > ... You had to downgrade your local ubuntu version to get it working?
> 
> No no, just the one in the `pasta-skeleton` dockerfile

Huh, the current skeleton isn't working? Let me check that...

rralf commented:
Yikes, you're right, let me check that.

rralf commented:
Please find the fix here https://github.com/lfd/PaStA/commit/869f7eac58c8c94bd74b534b5066b7fce958d715

rsarky commented:
> > I am getting some weird issue where PaStA cannot parse the `default.cfg`
> 
> I guess you will have to rebase this series to master/next. E.g., Rohit, you just changed the format of mbox-result -> patch-groups. This might explain the break...

Hmm, the weird thing is the same error is coming even when I checkout the master branch. Will try to debug this.

Apart from this another issue that I will face is since I have symlinked the linux repository in resources/linux/repo to somewhere else on my system it wont be available in the docker container. 
Wondering if there's any way other than just cloning the entire repository again there

rralf commented:
Currently, I use docker only for integration testing. Other than that, you can modify / extend the docker configuration to mount your local directories.

rsarky commented:
> I am getting some weird issue where PaStA cannot parse the default.cfg

This was an error with the toml version. I was using version `0.9.x` incorrectly, `0.10.x` works as required

rsarky commented:
> Currently, I use docker only for integration testing. Other than that, you can modify / extend the docker configuration to mount your local directories.

Thanks! Successfully mounted Pasta and Patchwork to docker containers and am able to run Pasta's commands now in the container after mounting my linux source code directory too.
Will work on their interaction now

rralf commented:
> > I am getting some weird issue where PaStA cannot parse the default.cfg
> 
> This was an error with the toml version. I was using version `0.9.x` incorrectly, `0.10.x` works as required

Aah, good to know!

rralf commented:
> > Currently, I use docker only for integration testing. Other than that, you can modify / extend the docker configuration to mount your local directories.
> 
> Thanks! Successfully mounted Pasta and Patchwork to docker containers and am able to run Pasta's commands now in the container after mounting my linux source code directory too.
> Will work on their interaction now

Did you write your own docker file for that? I'm willing to accept it, as this is clearly a better way for developing on PaStA.

rsarky commented:
> > > Currently, I use docker only for integration testing. Other than that, you can modify / extend the docker configuration to mount your local directories.
> > 
> > 
> > Thanks! Successfully mounted Pasta and Patchwork to docker containers and am able to run Pasta's commands now in the container after mounting my linux source code directory too.
> > Will work on their interaction now
> 
> Did you write your own docker file for that? I'm willing to accept it, as this is clearly a better way for developing on PaStA.

Didnt need to write my own docker file.
I use ` docker run -it --rm --network patchwork_default --name pasta -v <host/path/to/PaStA>:/home/pasta -v <host/path/to/linux-repo>:/home/linux pasta:latest`

And then symlink the linux repo in `resources/linux/repo` to `/home/linux/` or depending on where you have cloned the repository you can change the paths.

Wonder why the wrong toml version gets downloaded though. Whenever I bring the docker container up I have to upgrade the pip toml package to make PaStA work.

metp commented:
> Wonder why the wrong toml version gets downloaded though. Whenever I bring the docker container up I have to upgrade the pip toml package to make PaStA work.

Does it work with `RUN apt update && apt install -y --no-install-recommends` instead of just `RUN apt install -y --no-install-recommends` in the skeleton Dockerfile (without using docker build cache `--no-cache`)? The local package index is possibly outdated in the ubuntu image. Fetching an updated version of the ubuntu base image could work as well `docker build --pull .`

rsarky commented:
> > Wonder why the wrong toml version gets downloaded though. Whenever I bring the docker container up I have to upgrade the pip toml package to make PaStA work.
> 
> Does it work with `RUN apt update && apt install -y --no-install-recommends` instead of just `RUN apt install -y --no-install-recommends` in the skeleton Dockerfile (without using docker build cache `--no-cache`)? The local package index is possibly outdated in the ubuntu image. Fetching an updated version of the ubuntu base image could work as well `docker build --pull .`

Ended up using `20.04` instead of `18.04` version for ubntu and the correct package version is fetched.

rralf commented:
Yeah, I already pushed the upgrade to 20.04 to the next branch.

rsarky commented:
> Yeah, I already pushed the upgrade to 20.04 to the next branch.

Yup that's the one I am using now! Thanks

rsarky commented:
Have been slightly busy this week. @Honeybyte Another thing I was wondering was, how do you feed mails into patchwork? Did you use a downloaded mailbox. I am assuming you alse made Patchwork parse incoming mails with an existing email address?

metp commented:
> Have been slightly busy this week. @Honeybyte Another thing I was wondering was, how do you feed mails into patchwork? Did you use a downloaded mailbox. I am assuming you alse made Patchwork parse incoming mails with an existing email address?

Hi @rsarky,
"Official way": https://patchwork.readthedocs.io/en/latest/development/installation/#import-mailing-list-archives. However this only works for mailing lists archives that use pipermail (e.g those on https://lists.ozlabs.org/listinfo/). https://www.kernel.org/lore.html uses public inboxes (pubins). You can write a bash script in case you want to import those (don't have mine anymore 😞). Either clone your desired pubin and let the script iterate trough the git repository. Every commit in pubin relates to one mail. So you could possibly create your own mbox file and import it with https://patchwork.readthedocs.io/en/latest/deployment/management/#parsearchive. Or you could try to use pubin's [export to mbox](https://lore.kernel.org/lkml/_/text/help/) option. Unfortunately only single threads are downloadable.

rralf commented:
The next question will then be how we can synchronise PaStA's patch database with Patchwork. Back then, we already had some discussions on that.

My standpoint is that we can do the analysis detached from patchwork. The translate Message-IDs to patchwork IDs and establish the connection. My reasons: Public Inboxes are a great exchange format, and they're already present. We don't need invasive changes of PaStA.

@bulwahn 's standpoint is that we should use official data from patchwork. This requires us to regularly update our mail database from patchwork, but has the advantage that we can directly work on patchwork ids and won't run out of sync.

Whatever we choose, we always have the problem that clusters may change over time... I don't know into which problems we will run on patchwork side.

bulwahn commented:
Agree, with all said above. Here my summary:

There are two use cases needed and described:

1. First Use Case (Development Use Case for Testing the Integration)
To actually test the patchwork and pasta integration at scale, the developers need to set up a patchwork instance that simulates the actual use wrt. a large database that the users' patchwork instance would run with. Hence, we need quickly fill the patchwork database with emails. Mete explained how we use the existing archives to turn them into mbox format and then import them into our development&testing patchwork instance.

2. Second Use Case (The Patchwork Instance Administrator's Use Case)
We actually need to assume that the patchwork instance administrator has already set up their patchwork instance just as they wanted it to work, e.g., it could be following multiple public mailing lists (which might be archived or not), it could be just importing from a private mailbox etc.
So, they are relunctant to change their patchwork instance import just to make use of pasta. Hence,  I suggested to ensure that all emails are exported from the patchwork instance and handed over to pasta. Mete implemented that by having a script for the initial export of the existing mails in the instance, and then having another cronjob-suitable script for continuous export of new emails. 

Ralf envisioned that patchwork would take public inbox repositories as sole input. I share that vision, but it requires significant changes in patchwork itself, and it requires much broader acceptance and adoption of public inbox (so it remains a long-term goal we should work towards), but it cannot be part of this first short-term solution.

metp commented:
Even though my implementation was using Lukas' suggested approach I would rather stick to Ralf's now as Patchwork recently allows to [search for msgids via api](https://github.com/getpatchwork/patchwork/commit/d08b6c72964898c9997a62e4ab6a721f166a56ca). Lukas, your first use case is implementation independent from PaStA, the second one is also now satisfied when using pubins. It only requires updating Patchwork to 2.2.

Edit:
Example: http://patchwork.ozlabs.org/api/patches/?msgid=200809050416.27831.adetsch%40br.ibm.com

rsarky commented:
> Hi @rsarky,
> "Official way": https://patchwork.readthedocs.io/en/latest/development/installation/#import-mailing-list-archives. However this only works for mailing lists archives that use pipermail (e.g those on https://lists.ozlabs.org/listinfo/). https://www.kernel.org/lore.html uses public inboxes (pubins). You can write a bash script in case you want to import those (don't have mine anymore ). Either clone your desired pubin and let the script iterate trough the git repository. Every commit in pubin relates to one mail. So you could possibly create your own mbox file and import it with https://patchwork.readthedocs.io/en/latest/deployment/management/#parsearchive. Or you could try to use pubin's [export to mbox](https://lore.kernel.org/lkml/_/text/help/) option. Unfortunately only single threads are downloadable.

Thanks a lot for this @Honeybyte 
So currently what I have done is on say the [alsa-devel](https://lore.kernel.org/alsa-devel/) archive I wanted to download the archive in mbox format. Instead of writing a script for converting the pubin repo into one, I kind of found a shorter way for now. I searched for all mails using the query [https://lore.kernel.org/alsa-devel/?q=s%3A*](https://lore.kernel.org/alsa-devel/?q=s%3A*) in the archive. This gives an option to download the results in mbox form. 
Although I do think a script converting pubin repos to mboxes will be useful. Maybe it can be a binary in PaStA? As we have data structures for PubIn in Mbox.py in pasta.

Edit: imported patches into patchwork:
![1588083441_1947_28 04 2020_1366x768](https://user-images.githubusercontent.com/24373707/80498568-81d58380-8989-11ea-8cdd-0882b90f0136.png)

bulwahn commented:
> Even though my implementation was using Lukas' suggested approach I would rather stick to Ralf's now as Patchwork recently allows to [search for msgids via api](https://github.com/getpatchwork/patchwork/commit/d08b6c72964898c9997a62e4ab6a721f166a56ca). Lukas, your first use case is implementation independent from PaStA, the second one is also now satisfied when using pubins. It only requires updating Patchwork to 2.2.
> 

Okay, I got that. So if we want to support private mailboxes, where no public inbox repository exists. we would need to still need to create a public inbox repository somehow, either directly from the email account or by exporting from patchwork, and then we create a public inbox repository and provide that to pasta. Right?

rsarky commented:
@Honeybyte Did you also make some changes to some network configuration for docker on patchwork? Getting an error while connecting to port 8000 from PaStA's container. (`connection refused`) Connecting from to port 8000 from host works though

Edit: Ignore. Fixed this. The fix was using the `patchwork_default` container's IP address

metp commented:
> > Even though my implementation was using Lukas' suggested approach I would rather stick to Ralf's now as Patchwork recently allows to [search for msgids via api](https://github.com/getpatchwork/patchwork/commit/d08b6c72964898c9997a62e4ab6a721f166a56ca). Lukas, your first use case is implementation independent from PaStA, the second one is also now satisfied when using pubins. It only requires updating Patchwork to 2.2.
> 
> Okay, I got that. So if we want to support private mailboxes, where no public inbox repository exists. we would need to still need to create a public inbox repository somehow, either directly from the email account or by exporting from patchwork, and then we create a public inbox repository and provide that to pasta. Right?

Yes. Hmm unfortunately not a nice routine...

rsarky commented:
I reckon this can be closed now, @rralf ?

rralf commented:
Yes, thanks.
